### PR TITLE
Blacklist problematic UGER nodes

### PR DESCRIPTION
**Pull request by haydenm**
To bypass transient issues on a set of UGER nodes, maintain a
list of nodes that should be blacklisted. This list is maintained
by jobscript.sh and used in calls to qsub by cluster-submitter.py.

### CODE REVIEW

File: pipes/Broad_UGER/jobscript.sh [hunk 1]
```diff
@@ -15,11 +15,30 @@ export PATH="$MINICONDADIR/bin:$PATH"
...
-# since the node with the job doesn't have the NFS share mounted.
-# Hopefully it will be sent to a node that has it mounted.
-# For this to work, re-run must be set to y via "qsub -r y"
-ls "$DATADIR" || exit 99
+
+# As a simple solution to transient UGER problems, maintain a list of blacklisted
+# nodes. When a node fails a check, add its hostname to the list and exit 99
+# to reschedule the job. Blacklist these nodes via qsub
+# Maintain the list of blacklisted nodes as filenames in /broad/hptmp/$(whoami)/blacklisted-nodes
+BLACKLISTED_NODES="/broad/hptmp/$(whoami)/blacklisted-nodes/"
```
Comment (yesimon) [inline pipes/Broad_UGER/jobscript.sh]: I would recommend leaving final slash out as convention

File: pipes/Broad_UGER/cluster-submitter.py [hunk 2]
```diff
@@ -27,9 +37,17 @@
 if mem:
     # on UGER, memory requests are per-core (according to BITS as of Sept. 6, 2016)
     mem_per_core = round((int(mem)*1024)/int(cores), 2)
-    cmdline += ' -l h_vmem={}M,h_rss={}M '.format(  mem_per_core, round(1.2 * mem_per_core, 2)  )
+    if blacklisted_nodes:
+        cmdline += " -l h_vmem={}M,h_rss={}M,h='!({})' ".format(
```
Comment (tomkinsc) [inline pipes/Broad_UGER/cluster-submitter.py]: We should probably have comment lines here explaining the various resource parameters, or at least for `h=` mention that it's the hostname param, that it [accepts a regex](http://wiki.gridengine.info/wiki/index.php/EvaluationExpressionSupport), and we take advantage of that invert the match and blacklist hostnames.

**Comment by tomkinsc** [comment]
Let's hold off on merging this until we see if we can modify snakemake to avoid semaphore-based locks. It would still be helpful to have the blacklist ability, but we may be able to get away without checking the available shared memory.

**Comment by dpark01** [comment]
If you end up solving it that way, then perhaps it's good anyway to add a manually-curated blacklist (that is read-only as far as our scripts are concerned). Or maybe you just maintain it based on swappyness.

**Comment by haydenm** [comment]
I think we should maintain this list dynamically in its current framework. It's an easy way to fix various other node-specific issues that could (and likely will) arise in the future. But it's also a list we can manually add to and delete from. Perhaps we can add a flag to the top of `jobscript.sh` that specifies whether we want to check shared memory (and the other checks)? Of course, if the modification of snakemake solves it, we would set that flag to false.

**Comment by tomkinsc** [comment]
A [PR is open](https://bitbucket.org/snakemake/snakemake/pull-requests/109/add-force-use-threads-option-to-force/diff) to add a `--force-use-threads` option to Snakemake. If/when it is accepted, we can add the flag to our Snakemake call [in `run-pipe.sh`](https://github.com/broadinstitute/viral-ngs/blob/master/pipes/Broad_UGER/run-pipe.sh#L25-L32) once we update to a new release of Snakemake that includes the feature.

**Comment by tomkinsc** [comment]
Maybe change the [conda config](https://github.com/broadinstitute/viral-ngs/blob/master/travis/install-conda.sh#L43) to also depend on the `conda-forge` channel:
```shell
conda config --add channels bioconda
conda config --add channels conda-forge # <===
conda config --add channels r
```
The [docs](https://github.com/broadinstitute/viral-ngs/blob/master/docs/install.rst#configure-conda), and the easy-deploy script should also be updated if we add another channel dependency. In the script, multiple channels can be specified with additional `-c` params: `-c <channel> -c <channel2>`

**Pull request merged**